### PR TITLE
2origin

### PR DESCRIPTION
**Pull request by pbalu-code**
Centos 8 and Ubuntu 18 support added.
Tested in official vagrant images.
Centos 8 uses mariadb's repo.
Ubuntu uses a mirror.
Python 3 compatibility. (New Python modules...)

### CODE REVIEW

**Comment by bertvv** [comment]
Thank you for adding support for Ubuntu and EL8. However, I won't be able to merge this as is. I already implemented support for EL8 (with a different strategy than you) before I reviewed this PR and merged a few other PRs. So, now there's too many merge conflicts.

I don't plan to support Python 2 installations and prefer that it's assumed that Python 3 is present on the system.

Would you consider to reimplement Ubuntu support (preferrably 20.04 rather than 18.04) based on the latest changes in trunk? If not, I totally understand and I will try to make the necessary adjustments myself (giving you credit of course!).

Thanks again for taking the time and the effort you put into this!

File: defaults/main.yml [hunk 1]
```diff
@@ -3,11 +3,18 @@
 
 mariadb_databases: []
 mariadb_users: []
-mariadb_root_password: ''
+mariadb_root_password: root
```
Comment (bertvv) [inline defaults/main.yml:6]: Is it necessary to have a non-empty default value? The default installation leaves the root password blank, I would think it's best to stay consistent with that.

File: defaults/main.yml [hunk 2]
```diff
@@ -3,11 +3,18 @@
...
 
 mariadb_mirror: yum.mariadb.org
 mariadb_version: '10.4'
 
+mariadb_apt_mirror: 'http://mirror.vpsfree.cz'
+mariadb_apt_repositories: "deb [arch=amd64,arm64,ppc64el] {{ mariadb_apt_mirror }}/mariadb/repo/{{ mariadb_version }}/{{ ansible_distribution | lower }} {{ ansible_distribution_release }} main"
```
Comment (bertvv) [inline defaults/main.yml:12]: Since these variables are specific to Apt, I think these should go into `vars/ubuntu.yml`

File: tasks/configubuntu.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# roles/mariadb/tasks/configubuntu.yml
```
Comment (bertvv) [inline tasks/configubuntu.yml:1]: This file introduces a lot of code duplication. Is this necessary? What's the difference between this file and the original `config.yml`? Maybe adding `when: ansible_os_family == "RedHat"` or `when: ansible_os_family == "Debian"` to the necessary tasks is sufficient?

File: tasks/databases.yml [hunk 4]
```diff
@@ -20,6 +20,7 @@
   with_items: "{{ mariadb_databases }}"
   register: db_creation
   tags: mariadb
+#  debugger: always
```
Comment (bertvv) [inline tasks/databases.yml:23]: You forgot to remove these comments ;-)

File: tasks/install.yml [hunk 5]
```diff
@@ -1,18 +1,72 @@
...
+    - ansible_os_family == "RedHat"
+    - ansible_facts['distribution_major_version'] == "7" or ansible_facts['distribution_major_version'] == "30"
+
+- name: Check MariaDB Centos/RegHat 8 repo installed
+  stat:
+    path: /etc/yum.repos.d/mariadb.repo
+  register: c8Mariadbpreo
 
-- name: Install packages
+- name: Install Centos/RedHat 8 MariaDB repo
```
Comment (bertvv) [inline tasks/install.yml:21]: I now install the .repo file using a template (the `yum_repository:` task causes an issue with EL8. The change you propose, downloading and executing a script, seems more kludgy.

File: tasks/install.yml [hunk 6]
```diff
@@ -1,18 +1,72 @@
...
+    - ansible_facts['distribution_major_version'] == "8"
+
+- name: Install Install MariaDB
   package:
     name: "{{ item }}"
     state: installed
   with_items: "{{ mariadb_packages }}"
   tags: mariadb
+
+- name: Install MySQL-python Centos/RedHat 7
```
Comment (bertvv) [inline tasks/install.yml:46]: I think that this task and the following can now be more efficiently implemented using appropriate `vars/` files. I changed the `include_vars:` task at the beginning of `main.yml` to also allow us to specify versions.

File: tasks/installubuntu.yml [hunk 7]
```diff
@@ -0,0 +1,47 @@
+# roles/mariadb/tasks/install.yml
+---
+- name: Check Python3 present
```
Comment (bertvv) [inline tasks/installubuntu.yml:3]: I'd rather not have to add code that checks Python versions, that seems outside the scope of this role.

File: tasks/installubuntu.yml [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  stat:
+    path: /usr/bin/python3
+  register: python3ok
+
+- name: correct python version selected
```
Comment (bertvv) [inline tasks/installubuntu.yml:8]: I fear this task could have side effects beyond the execution of this role. Selecting a system-wide Python version may affect other components of the system than the MariaDB installation.

File: tasks/root-password.yml [hunk 9]
```diff
@@ -7,7 +7,8 @@
       Warning!! the MariaDB root password was left empty. Please set a custom
       password with role variable mariadb_root_password to secure your database
       server!
-  when: mariadb_root_password | length == 0
+#  when: mariadb_root_password | length == 0   ---It is not Vault compatible
+  when: mariadb_root_password == ''
```
Comment (bertvv) [inline tasks/root-password.yml:11]: In the current version, this check is implemented as `when: not mariadb_root_password`. Would that be Vault compatible?

File: tasks/root-password.yml [hunk 10]
```diff
@@ -19,10 +20,12 @@
   tags: mariadb
 
 - name: Set MariaDB root password for the first time (root@localhost)
+#  become_user: mysql
```
Comment (bertvv) [inline tasks/root-password.yml:23]: Should this comment line (and the one below) be removed or uncommented?

File: defaults/main.yml (hunk 2 above)
Comment (pbalu-code) [reply defaults/main.yml:12]: You right! :-)

File: defaults/main.yml (hunk 1 above)
Comment (pbalu-code) [reply defaults/main.yml:6]: The empty password dangerous. Except learning processes and some docker/openshift application we should set password. This is only my opinion of course, not the official suggestion.

**Comment by pbalu-code** [comment]
Hello!
Thank you for your time and  for the comments.
I have started to work on it. I'll re-implement the Ubuntu install section soon to support 20.04 and I'll send a new pull request. (That time when I worked on it, the vagrant image for Ubuntu 20.04 was wrong)

**Comment by bertvv** [comment]
Thank you! Adding support for Ubuntu would be awesome. I never got to this myself...
